### PR TITLE
Elaborate on WebSocket multiple sends/receives

### LINKED ISSUE #8040: Clarify `WebSocket.SendAsync()` limitations and API guarantees
The `WebSocket.SendAsync()` [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.net.websockets.websocket.sendasync?view=net-6.0&source=docs) states that:

> Exactly one send and one receive is supported on each [WebSocket](https://docs.microsoft.com/en-us/dotnet/api/system.net.websockets.websocket?view=net-6.0) object in parallel.

However it seems that calling `SendAsync()` multiple times before the previous send completed does not raise any exception, so this statement seems to refer to the internal implementation (the sending is done serially), which obviously comes from a limitation of the WebSocket protocol itself, rather than a limitation in the public API. So the first ask would be to clarify this in the docs, and explain that calling this method multiple times is allowed, and what happens in that case (likely: sends are queued internally).

The second part is, assuming those calls are internally queued, what guarantees does the API provide? Specifically, is the API guaranteeing that sends are executed in the same order as the calls are made? This is critical for some scenario, and currently I found myself having to work around the lack of API guarantee by maintaining my own queue (and therefore, having to copy the data to send into that manual queue), and making calls to `SendAsync()` one by one, waiting for the previous call to complete before I make the next one. This is overly complicated, results in extra copy, and I'm pretty confident is completely useless as I would expect the actual implementation to be doing exactly the same thing internally already (duplicated work).

In summary, this issue is 2 related asks:
1. Clarify whether `SendAsync()` can be called multiple times while a send is pending. This is somewhat related to #7809.
2. If so, clarify the behavior of the implementation in terms of internally enqueuing those send calls, and the ordering guarantee if any.

### PR DESCRIPTION
**Pull request by CarnaViire**
## Summary

Clarify that multiple sends and multiple receives on a WebSocket are not supported.

Fixes #8040

### CODE REVIEW

**Comment by CarnaViire** [comment]
PTAL @dotnet/ncl @gewarren

**Review by liveans — APPROVED** [review]
Well, it looks good to me, but it'd be nice if somebody else can review this one as well.

File: xml/System.Net.WebSockets/ClientWebSocket.xml [hunk 1]
```diff
@@ -52,6 +52,8 @@
 ## Remarks
  Some of the classes and class elements in the <xref:System.Net.WebSockets> namespace are supported on Windows 7, Windows Vista SP2, and Windows Server 2008. However, the only public implementations of client and server WebSockets are supported on Windows 8 and Windows Server 2012. The class elements in the <xref:System.Net.WebSockets> namespace that are supported on Windows 7, Windows Vista SP2, and Windows Server 2008 are abstract class elements. This allows an application developer to inherit and extend these abstract class classes and class elements with an actual implementation of client WebSockets.
 
+ Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel. Issuing multiple sends or multiple receives at the same time (for example, without awaiting, or from multiple threads without synchronization) is *not supported* and will result in an *undefined behavior*. Users should ensure the previous operation is awaited (or completed) before issuing the next one. It is up to the user to serialize the access via whatever mechanism works best for them, for example, by using a lock or a semaphore.
```
Comment (gewarren) [inline xml/System.Net.WebSockets/ClientWebSocket.xml:55]:
```suggestion
Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel. Issuing multiple sends or multiple receives at the same time (for example, without awaiting, or from multiple threads without synchronization) is *not supported* and will result in an *undefined behavior*. Ensure that the previous operation is awaited (or completed) before issuing the next one. You should serialize the access via whatever mechanism works best for you, for example, by using a lock or a semaphore.
```

File: xml/System.Net.WebSockets/ClientWebSocket.xml [hunk 2]
```diff
@@ -588,7 +590,7 @@
 ## Remarks
  This operation will not block. The returned <xref:System.Threading.Tasks.Task%601> object will complete after the receive request on the <xref:System.Net.WebSockets.ClientWebSocket> instance has completed.
 
- Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel.
+ Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel. Issuing multiple receives at the same time is *not supported* and will result in an *undefined behavior*. Users should serialize receive operations via whatever mechanism works best for them, for example, by using a lock or a semaphore.
```
Comment (gewarren) [inline xml/System.Net.WebSockets/ClientWebSocket.xml:593]:
```suggestion
Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel. Issuing multiple receives at the same time is *not supported* and will result in an *undefined behavior*. You should serialize receive operations via whatever mechanism works best for you, for example, by using a lock or a semaphore.
```

File: xml/System.Net.WebSockets/ClientWebSocket.xml [hunk 3]
```diff
@@ -634,7 +636,7 @@
...
```
Comment (gewarren) [inline xml/System.Net.WebSockets/ClientWebSocket.xml:639]:
```suggestion
Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel. Issuing multiple receives at the same time is *not supported* and will result in an *undefined behavior*. You should serialize receive operations via whatever mechanism works best for you, for example, by using a lock or a semaphore.
```

File: xml/System.Net.WebSockets/ClientWebSocket.xml [hunk 4]
```diff
@@ -694,7 +696,7 @@
 
  This operation will not block. The returned <xref:System.Threading.Tasks.Task> object will complete after the send request on the <xref:System.Net.WebSockets.ClientWebSocket> instance has completed.
 
- Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel.
+ Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel. Issuing multiple sends at the same time is *not supported* and will result in an *undefined behavior*. Users should serialize send operations via whatever mechanism works best for them, for example, by using a lock or a semaphore.
```
Comment (gewarren) [inline xml/System.Net.WebSockets/ClientWebSocket.xml:699]:
```suggestion
Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel. Issuing multiple sends at the same time is *not supported* and will result in an *undefined behavior*. You should serialize send operations via whatever mechanism works best for you, for example, by using a lock or a semaphore.
```

File: xml/System.Net.WebSockets/ClientWebSocket.xml [hunk 5]
```diff
@@ -746,7 +748,7 @@
 
 This operation will not block. The returned <xref:System.Threading.Tasks.Task> object will complete after the send request on the <xref:System.Net.WebSockets.ClientWebSocket> instance has completed.
 
-Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel.
+Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel. Issuing multiple sends at the same time is *not supported* and will result in an *undefined behavior*. Users should serialize send operations via whatever mechanism works best for them, for example, by using a lock or a semaphore.
```
Comment (gewarren) [inline xml/System.Net.WebSockets/ClientWebSocket.xml:751]:
```suggestion
Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.ClientWebSocket> object in parallel. Issuing multiple sends at the same time is *not supported* and will result in an *undefined behavior*. You should serialize send operations via whatever mechanism works best for you, for example, by using a lock or a semaphore.
```

File: xml/System.Net.WebSockets/WebSocket.xml [hunk 6]
```diff
@@ -44,7 +44,9 @@
...
+ Some of the classes in the <xref:System.Net.WebSockets> namespace are supported on Windows 7, Windows Vista SP2, and Windows Server 2008. However, the only public implementations of client and server WebSockets are supported on Windows 8 and Windows Server 2012. The classes and class elements in the <xref:System.Net.WebSockets> namespace that are supported on Windows 7, Windows Vista SP2, and Windows Server 2008 are abstract classes. This allows an application developer to inherit and extend these abstract classes with an actual implementation of client WebSockets.
+
+ Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.WebSocket> object in parallel. Issuing multiple sends or multiple receives at the same time (for example, without awaiting, or from multiple threads without synchronization) is *not supported* and will result in an *undefined behavior*. Users should ensure the previous operation is awaited (or completed) before issuing the next one. It is up to the user to serialize the access via whatever mechanism works best for them, for example, by using a lock or a semaphore.
```
Comment (gewarren) [inline xml/System.Net.WebSockets/WebSocket.xml:49]:
```suggestion
Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.WebSocket> object in parallel. Issuing multiple sends or multiple receives at the same time (for example, without awaiting, or from multiple threads without synchronization) is *not supported* and will result in an *undefined behavior*. Ensure that the previous operation is awaited (or completed) before issuing the next one. Serialize the access via whatever mechanism works best for you, for example, by using a lock or a semaphore.
```

File: xml/System.Net.WebSockets/WebSocket.xml [hunk 7]
```diff
@@ -767,7 +769,7 @@
 ## Remarks  
  This operation will not block. The returned <xref:System.Threading.Tasks.Task%601> object will complete after the data has been received on the <xref:System.Net.WebSockets.WebSocket>.  
   
- Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.WebSocket> object in parallel.  
+ Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.WebSocket> object in parallel. Issuing multiple receives at the same time is *not supported* and will result in an *undefined behavior*. Users should serialize receive operations via whatever mechanism works best for them, for example, by using a lock or a semaphore.
```
Comment (gewarren) [inline xml/System.Net.WebSockets/WebSocket.xml:772]:
```suggestion
Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.WebSocket> object in parallel. Issuing multiple receives at the same time is *not supported* and will result in an *undefined behavior*. You should serialize receive operations via whatever mechanism works best for you, for example, by using a lock or a semaphore.
```

File: xml/System.Net.WebSockets/WebSocket.xml [hunk 8]
```diff
@@ -909,7 +911,7 @@
 ## Remarks  
  This operation will not block. The returned <xref:System.Threading.Tasks.Task> object will complete after the data has been sent on the <xref:System.Net.WebSockets.WebSocket>.  
   
- Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.WebSocket> object in parallel.  
+ Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.WebSocket> object in parallel. Issuing multiple sends at the same time is *not supported* and will result in an *undefined behavior*. Users should serialize send operations via whatever mechanism works best for them, for example, by using a lock or a semaphore.
```
Comment (gewarren) [inline xml/System.Net.WebSockets/WebSocket.xml:914]:
```suggestion
Exactly one send and one receive is supported on each <xref:System.Net.WebSockets.WebSocket> object in parallel. Issuing multiple sends at the same time is *not supported* and will result in an *undefined behavior*. You should serialize send operations via whatever mechanism works best for you, for example, by using a lock or a semaphore.  
```

**Review by gewarren — APPROVED** [review]
Some slight wording changes in voice - MS style is to talk directly to the reader/user, not in the third person.

**Pull request merged**